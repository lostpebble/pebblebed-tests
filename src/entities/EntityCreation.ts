import { sampleSize, random } from "lodash";
import { ICoordinates } from "../dataTypes/dataTypes";

export interface IDSTestEntityBase {
  date?: Date;
  tags: string[];
  amount?: number;
  location?: ICoordinates;
  worthy?: boolean;
  object?: object;
  testSerialization?: {
    time: Date;
    [key: string]: any;
  };
  numberArray?: number[];
  deepNumberArray?: { num: number; }[];
}

const tags = ["egg", "salad", "green", "blue", "ham", "great", "dont"];
const trueOrFalse = () => random(1) === 1;

export function createNewTestEntityId(withAll: boolean = false): IDSTestEntityBase {
  const numberArray = [123123, 222.233, 99595, 2220031];
  const deepNumberArray = [{ num: 23123.22 }, { num: 239992 }];

  return {
    amount: random(999999, true),
    date: withAll || trueOrFalse() ? new Date() : undefined,
    tags: sampleSize(tags, random(1, 5)),
    location: withAll || trueOrFalse()
      ? {
        longitude: random(-90, 90, true),
        latitude: random(-90, 90, true),
      }
      : undefined,
    worthy: withAll || trueOrFalse() ? trueOrFalse() : undefined,
    object: withAll || trueOrFalse() ? { test: "something" } : undefined,
    testSerialization: withAll || trueOrFalse() ? { time: new Date(), serialized: "serialized-test-thing" } : undefined,
    numberArray,
    deepNumberArray,
  };
}
