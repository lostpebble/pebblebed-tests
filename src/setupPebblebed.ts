import * as datastoreLib from "@google-cloud/datastore";
// import { Pebblebed } from "../../src";
import { Pebblebed } from "pebblebed";

const datastore = datastoreLib();

Pebblebed.connectDatastore(datastore);
Pebblebed.setDefaultNamespace("pebblebed-test-namespace");
