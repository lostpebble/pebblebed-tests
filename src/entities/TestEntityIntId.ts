import { Pebblebed, types } from "pebblebed";
import { DefaultDateTimeNow, ICoordinates } from "../dataTypes/dataTypes";
import { IDSTestEntityBase } from "./EntityCreation";
import { reviveDateObjects } from "../utility";

export interface IDSTestEntityIntId extends IDSTestEntityBase{
  idThing?: string;
}

const schema = Pebblebed.createSchema<IDSTestEntityIntId>({
  idThing: types.integerId(),
  amount: types.double(),
  date: DefaultDateTimeNow,
  location: types.geoPoint(),
  tags: types.array(),
  worthy: types.boolean({
    indexed: false,
  }),
  object: types.object({
    indexed: false,
  }),
  testSerialization: types.serializedJson({
    indexed: false,
    reviver: reviveDateObjects,
  }),
  deepNumberArray: types.array({ indexed: false }),
  numberArray: types.array({ indexed: false }),
}).setDefaultMeta({
  nullValueIfUnset: false,
});

export const TestEntityIntIdModel = Pebblebed.createModel("TestEntityIntId", schema);
