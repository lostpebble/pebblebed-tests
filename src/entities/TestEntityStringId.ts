import { Pebblebed, types, } from "pebblebed";
import { DefaultDateTimeNow } from "../dataTypes/dataTypes";
import { IDSTestEntityBase } from "./EntityCreation";

export interface IDSTestEntityStringId extends IDSTestEntityBase {
  idThing: string;
  // date?: Date;
  // tags: string[];
  // amount?: number;
  // location?: ICoordinates;
  // worthy?: boolean;
}

const schema = Pebblebed.createSchema<IDSTestEntityStringId>({
  idThing: types.stringId(),
  amount: types.double({
    indexed: false,
  }),
  date: DefaultDateTimeNow,
  location: types.geoPoint(),
  tags: types.array().required(),
  worthy: types.boolean(),
});

// export const TestEntityStringIdModel = new PebblebedModel("TestEntityStringId", schema);
export const TestEntityStringIdModel = Pebblebed.createModel("TestEntityStringId", schema);
