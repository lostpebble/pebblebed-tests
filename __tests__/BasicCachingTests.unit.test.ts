import "../src/setupPebblebed";
import {
  IDSTestEntityIntId,
  TestEntityIntIdModel
} from "../src/entities/TestEntityIntId";
import { createNewTestEntityId } from "../src/entities/EntityCreation";
import IoRedisLib from "ioredis";
import { Pebblebed, PebblebedDefaultRedisCacheStore } from "pebblebed";
import { waitSeconds } from "../src/utility";

describe(`It should cache entities`, () => {
  beforeAll(() => {
    jest.setTimeout(20000);
    const redis = new IoRedisLib();
    Pebblebed.setCacheStore(new PebblebedDefaultRedisCacheStore(redis));
  });

  it("Should create, save and load the same exact entity (basic)", async cb => {
    const newEntity: IDSTestEntityIntId = {
      date: new Date(),
      tags: ["egg", "salad"],
      amount: 2032.2323,
    };

    const {
      savedEntities: [savedEntity],
      generatedIds: [genId],
    } = await TestEntityIntIdModel.save(newEntity)
      .enableCaching(true)
      .cachingSeconds(100)
      .returnSavedEntities()
      .generateUnsetIds()
      .run();

    newEntity.idThing = genId;
    newEntity[Pebblebed.ds.KEY] = savedEntity[Pebblebed.ds.KEY];

    expect(savedEntity).toEqual(newEntity);

    const cachedEntity = await TestEntityIntIdModel.load(genId).first().run();

    await waitSeconds(0.5);
    const loadedNoCacheEntity = await TestEntityIntIdModel.load(genId).enableCaching(false).first().run();

    expect(cachedEntity).toEqual(savedEntity);
    expect(cachedEntity).toEqual(loadedNoCacheEntity);

    delete cachedEntity.date;
    delete newEntity.date;
    delete loadedNoCacheEntity.date;
    delete savedEntity.date;
    delete newEntity.idThing;
    delete loadedNoCacheEntity.idThing;
    delete savedEntity.idThing;
    delete cachedEntity.idThing;
    delete newEntity[Pebblebed.ds.KEY];
    delete loadedNoCacheEntity[Pebblebed.ds.KEY];
    delete savedEntity[Pebblebed.ds.KEY];
    delete cachedEntity[Pebblebed.ds.KEY];

    expect({ newEntity, savedEntity, loadedNoCacheEntity }).toMatchSnapshot();

    await TestEntityIntIdModel.delete().idsOrKeys(genId).run();
    await waitSeconds(0.5);
    const afterDeleteEntity = await TestEntityIntIdModel.load(genId).enableCaching(true).first().run();

    expect(afterDeleteEntity).toEqual(null);

    cb();
  });

  it("Should create, save and load the same exact entity (with random props)", async cb => {
    const newEntity: IDSTestEntityIntId = {
      date: new Date(),
      tags: ["egg", "salad"],
      amount: 2032.2323,
      location: {
        latitude: 23.223,
        longitude: -33.2221,
      },
      object: { test: "something" },
      worthy: false,
      testSerialization:  {
        somethingElse: "serialized",
        time: new Date("2019-03-13T17:16:34.099Z"),
      },
    };

    const {
      savedEntities: [savedEntity],
      generatedIds: [genId],
    } = await TestEntityIntIdModel.save(newEntity)
      .enableCaching(true)
      .cachingSeconds(100)
      .returnSavedEntities()
      .generateUnsetIds()
      .run();

    const cachedEntity = await TestEntityIntIdModel.load(genId).first().run();

    await waitSeconds(0.5);
    const loadedNoCacheEntity = await TestEntityIntIdModel.load(genId).enableCaching(false).first().run();

    expect(cachedEntity).toEqual(savedEntity);
    expect(cachedEntity).toEqual(loadedNoCacheEntity);

    expect(newEntity.testSerialization).toEqual(cachedEntity.testSerialization);
    expect(newEntity.object).toEqual(cachedEntity.object);
    expect(newEntity.location).toEqual(cachedEntity.location);
    expect(newEntity.worthy).toEqual(cachedEntity.worthy);
    expect(newEntity.amount).toEqual(cachedEntity.amount);
    expect(newEntity.date).toEqual(cachedEntity.date);
    expect(newEntity.tags).toEqual(cachedEntity.tags);

    delete cachedEntity.date;
    delete newEntity.date;
    delete loadedNoCacheEntity.date;
    delete savedEntity.date;
    delete newEntity.idThing;
    delete loadedNoCacheEntity.idThing;
    delete savedEntity.idThing;
    delete cachedEntity.idThing;
    delete newEntity[Pebblebed.ds.KEY];
    delete loadedNoCacheEntity[Pebblebed.ds.KEY];
    delete savedEntity[Pebblebed.ds.KEY];
    delete cachedEntity[Pebblebed.ds.KEY];

    expect({ newEntity, savedEntity, loadedNoCacheEntity }).toMatchSnapshot();

    await TestEntityIntIdModel.delete().idsOrKeys(genId).run();
    await waitSeconds(0.5);
    const afterDeleteEntity = await TestEntityIntIdModel.load(genId).enableCaching(true).first().run();

    expect(afterDeleteEntity).toEqual(null);

    cb();
  });

  it("Should create, save and load the same exact entity (with deliberate undefined set)", async cb => {
    // const newEntity: IDSTestEntityIntId = createNewTestEntityId(true);
    const newEntity: IDSTestEntityIntId = {
      amount: 203222223.2322113,
      date: undefined,
      tags: undefined,
      location: undefined,
      object: undefined,
      worthy: undefined,
      testSerialization: undefined,
    };

    const {
      savedEntities: [savedEntity],
      generatedIds: [genId],
    } = await TestEntityIntIdModel.save(newEntity)
      .enableCaching(true)
      .cachingSeconds(100)
      .returnSavedEntities()
      .generateUnsetIds()
      .run();

    const cachedEntity = await TestEntityIntIdModel.load(genId).first().run();

    await waitSeconds(0.5);
    const loadedNoCacheEntity = await TestEntityIntIdModel.load(genId).enableCaching(false).first().run();

    delete newEntity.worthy;
    delete newEntity.object;
    delete newEntity.tags;
    newEntity.date = savedEntity.date;
    delete newEntity.location;

    expect(cachedEntity).toEqual(savedEntity);
    expect(cachedEntity).toEqual(loadedNoCacheEntity);

    expect(newEntity.testSerialization).toEqual(cachedEntity.testSerialization);
    expect(newEntity.object).toEqual(cachedEntity.object);
    expect(newEntity.location).toEqual(cachedEntity.location);
    expect(newEntity.worthy).toEqual(cachedEntity.worthy);
    expect(newEntity.amount).toEqual(cachedEntity.amount);
    expect(newEntity.date).toEqual(cachedEntity.date);
    expect(newEntity.tags).toEqual(cachedEntity.tags);

    await TestEntityIntIdModel.delete(savedEntity).run();
    await waitSeconds(0.5);
    const afterDeleteEntity = await TestEntityIntIdModel.load(genId).enableCaching(true).first().run();

    expect(afterDeleteEntity).toEqual(null);

    delete cachedEntity.date;
    delete newEntity.date;
    delete loadedNoCacheEntity.date;
    delete savedEntity.date;
    delete newEntity.idThing;
    delete loadedNoCacheEntity.idThing;
    delete savedEntity.idThing;
    delete cachedEntity.idThing;
    delete newEntity[Pebblebed.ds.KEY];
    delete loadedNoCacheEntity[Pebblebed.ds.KEY];
    delete savedEntity[Pebblebed.ds.KEY];
    delete cachedEntity[Pebblebed.ds.KEY];

    expect({ newEntity, savedEntity, loadedNoCacheEntity }).toMatchSnapshot();

    cb();
  });

  it("Should create, save and load the same exact entity (with deliberately set ID)", async cb => {
    // const newEntity: IDSTestEntityIntId = createNewTestEntityId(true);
    const newEntity: IDSTestEntityIntId = {
      idThing: "293939223",
      date: new Date(),
      tags: ["egg", "salad"],
      amount: 2032.2323,
      location: {
        latitude: 23.223,
        longitude: -33.2221,
      },
      object: { test: "something" },
      worthy: false,
      testSerialization:  {
        somethingElse: "serialized",
        time: new Date("2019-03-13T17:25:09.479Z"),
      },
    };

    const {
      savedEntities: [savedEntity],
    } = await TestEntityIntIdModel.save(newEntity)
      .enableCaching(true)
      .cachingSeconds(100)
      .returnSavedEntities()
      .run();

    const cachedEntity = await TestEntityIntIdModel.load(newEntity.idThing).first().run();

    await waitSeconds(0.5);
    const loadedNoCacheEntity = await TestEntityIntIdModel.load(newEntity.idThing).enableCaching(false).first().run();

    expect(cachedEntity).toEqual(savedEntity);
    expect(cachedEntity).toEqual(loadedNoCacheEntity);

    newEntity.idThing = savedEntity.idThing;
    await TestEntityIntIdModel.delete(newEntity).run();
    await waitSeconds(0.5);
    const afterDeleteEntity = await TestEntityIntIdModel.load(newEntity.idThing).enableCaching(true).first().run();

    expect(afterDeleteEntity).toEqual(null);

    expect(newEntity.testSerialization).toEqual(cachedEntity.testSerialization);
    expect(newEntity.object).toEqual(cachedEntity.object);
    expect(newEntity.location).toEqual(cachedEntity.location);
    expect(newEntity.worthy).toEqual(cachedEntity.worthy);
    expect(newEntity.amount).toEqual(cachedEntity.amount);
    expect(newEntity.date).toEqual(cachedEntity.date);
    expect(newEntity.tags).toEqual(cachedEntity.tags);

    delete cachedEntity.date;
    delete newEntity.date;
    delete loadedNoCacheEntity.date;
    delete savedEntity.date;

    expect({ newEntity, savedEntity, loadedNoCacheEntity }).toMatchSnapshot();

    cb();
  });
});
