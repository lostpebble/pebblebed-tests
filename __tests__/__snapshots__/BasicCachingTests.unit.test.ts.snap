// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should cache entities Should create, save and load the same exact entity (basic) 1`] = `
Object {
  "loadedNoCacheEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "tags": Array [
      "egg",
      "salad",
    ],
  },
  "newEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "tags": Array [
      "egg",
      "salad",
    ],
  },
  "savedEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "tags": Array [
      "egg",
      "salad",
    ],
  },
}
`;

exports[`It should cache entities Should create, save and load the same exact entity (with deliberate undefined set) 1`] = `
Object {
  "loadedNoCacheEntity": Object {
    "amount": 203222223.2322113,
  },
  "newEntity": Object {
    "amount": 203222223.2322113,
    "testSerialization": undefined,
  },
  "savedEntity": Object {
    "amount": 203222223.2322113,
  },
}
`;

exports[`It should cache entities Should create, save and load the same exact entity (with deliberately set ID) 1`] = `
Object {
  "loadedNoCacheEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "idThing": "293939223",
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "somethingDeeper": Object {
        "again": 222333,
        "great": false,
        "next": Array [
          10,
          9,
          8,
        ],
      },
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:25:09.479Z,
    },
    "worthy": false,
    Symbol(KEY): Key {
      "id": "293939223",
      "kind": "TestEntityIntId",
      "namespace": "pebblebed-test-namespace",
      "path": Array [
        "TestEntityIntId",
        "293939223",
      ],
    },
  },
  "newEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "idThing": "293939223",
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "somethingDeeper": Object {
        "again": 222333,
        "great": false,
        "next": Array [
          10,
          9,
          8,
        ],
      },
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:25:09.479Z,
    },
    "worthy": false,
  },
  "savedEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "idThing": "293939223",
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "somethingDeeper": Object {
        "again": 222333,
        "great": false,
        "next": Array [
          10,
          9,
          8,
        ],
      },
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:25:09.479Z,
    },
    "worthy": false,
    Symbol(KEY): Key {
      "id": "293939223",
      "kind": "TestEntityIntId",
      "namespace": "pebblebed-test-namespace",
      "path": Array [
        "TestEntityIntId",
        "293939223",
      ],
    },
  },
}
`;

exports[`It should cache entities Should create, save and load the same exact entity (with random props) 1`] = `
Object {
  "loadedNoCacheEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:16:34.099Z,
    },
    "worthy": false,
  },
  "newEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:16:34.099Z,
    },
    "worthy": false,
  },
  "savedEntity": Object {
    "amount": 2032.2323,
    "deepNumberArray": Array [
      Object {
        "num": 2322.22,
      },
      Object {
        "num": 222222222,
      },
    ],
    "location": Object {
      "latitude": 23.223,
      "longitude": -33.2221,
    },
    "numberArray": Array [
      2222.22,
      5123123,
      -99999.223,
    ],
    "object": Object {
      "test": "something",
    },
    "tags": Array [
      "egg",
      "salad",
    ],
    "testSerialization": Object {
      "somethingElse": "serialized",
      "time": 2019-03-13T17:16:34.099Z,
    },
    "worthy": false,
  },
}
`;
